pokedex_async_redux/lib/features/pokemon_overview/pokemon_overview_page.dart

```diff
@@ -1,68 +1,141 @@
+import 'dart:async';
+
 import 'package:flutter/material.dart';
 import 'package:pokedex_asyn_redux/api/model/pokemon.dart';
 import 'package:pokedex_asyn_redux/features/pokemon_details/pokemon_details_connector.dart';
 import 'package:pokedex_asyn_redux/features/pokemon_overview/widgets/pokemon_card.dart';
+import 'package:pokedex_asyn_redux/features/pokemon_overview/widgets/pokemon_search_bar.dart';
 import 'package:pokedex_asyn_redux/utils/async.dart';
 import 'package:pokedex_asyn_redux/utils/colors.dart';
+import 'package:pokedex_asyn_redux/utils/number_constants.dart';
 import 'package:pokedex_asyn_redux/utils/string_constants.dart';
 import 'package:pokedex_asyn_redux/widgets/reusable_appbar.dart';
 import 'package:pokedex_asyn_redux/widgets/spacing.dart';
 
-class PokemonOverviewPage extends StatelessWidget {
+class PokemonOverviewPage extends StatefulWidget {
   const PokemonOverviewPage({
     required this.pokemons,
+    required this.searchedPokemons,
+    required this.filterPokemons,
+    required this.clearSearchedPokemons,
     Key? key,
   }) : super(key: key);
 
   final Async<List<Pokemon>> pokemons;
+  final List<Pokemon> searchedPokemons;
+  final VoidCallback clearSearchedPokemons;
+  final Function(String) filterPokemons;
+
+  @override
+  State<PokemonOverviewPage> createState() => _PokemonOverviewPageState();
+}
+
+class _PokemonOverviewPageState extends State<PokemonOverviewPage> {
+  late TextEditingController _searchController;
+  late bool isSearching;
+  Timer? _debounceTimer;
+  late String _searchQuery = emptyString;
+
+  @override
+  void initState() {
+    super.initState();
+    isSearching = false;
+    _searchController = TextEditingController();
+  }
+
+  @override
+  void dispose() {
+    _searchController.dispose();
+    _debounceTimer?.cancel();
+    if (widget.searchedPokemons.isNotEmpty) {
+      widget.clearSearchedPokemons();
+    }
+    super.dispose();
+  }
 
   @override
   Widget build(BuildContext context) {
     return Scaffold(
       appBar: const CustomAppBar(title: pokemonOverviewTitle),
-      body: pokemons.when(
-        (data) {
-          return GridView.builder(
-            itemCount: data.length,
-            gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 2),
-            itemBuilder: (_, index) {
-              final pokemon = data[index];
-              return PokemonCard(
-                pokemon: pokemon,
-                onTap: () => _navigateToPokemonDetailsPage(context: context, pokemon: pokemon),
-              );
+      body: Column(
+        children: [
+          PokemonSearchBar(
+            textController: _searchController,
+            clearSearchBar: _clearSearchedPokemons,
+            textChanged: (val) {
```

Comment:
no need for this, maximize the addListener of your text controller 🙂 

like this:
`_searchController = TextEditingController()..addListener(_onSearchPokemons);`